Add URAT catalog information

geordie666 opened the pull request:
This PR adds [the URAT catalog](http://cdsarc.u-strasbg.fr/viz-bin/Cat?I/329) to desitarget. It includes:

- A new module `uratmatch.py` to retrieve URAT data from Vizier/FTP and reformat it.
- Code to efficiently match RAs/Decs to URAT, as part of the `uratmatch.py` module.
- Code to substitute URAT proper motions for GFAs where Gaia has not yet measured proper motions (the other Gaia information is retained for a GFA target, but zero/NaN proper motions are overwritten with URAT values, and the URAT_ID and separation is recorded).

This addresses #520. Here's an example of a general region we uncovered that had zero objects in a guide camera, which now has multiple objects from URAT:
<img width="853" alt="URATGFAs" src="https://user-images.githubusercontent.com/13952581/62398531-385d2280-b536-11e9-87be-7b6b8c9ec696.png">

sbailey commented on py/desitarget/fortran/README:
```diff
@@ -0,0 +1,17 @@
+
+v1dump: An executable compiled at NERSC via:
+
+        gfortran -o v1dump ADM-v1dump.f v1sub.f
```
Thanks for including specific instructions for compiling v1dump.  Please also include documentation somewhere about how you ran it and were the outputs are that most
users should just use without having to run this themselves.

geordie666 replied:
Done

sbailey commented on py/desitarget/uratmatch.py:
```diff
@@ -0,0 +1,612 @@
+    # ADM check that the $URAT_DIR environment variable is set.
+    uratdir = os.environ.get('URAT_DIR')
+    if uratdir is None:
+        msg = "Set $URAT_DIR environment variable!"
```
Let's give the user a hint of where that directory should be when running at NERSC.  Please add it to `etc/desitarget.module` so that our code installations at NERSC will automatically set it.

geordie666 replied:
Done.

sbailey commented on py/desitarget/fortran/README:
```diff
@@ -0,0 +1,17 @@
+v1sub.f: Routines to help parse UCAC/URAT files, taken directly from
+         http://cdsarc.u-strasbg.fr/ftp/I/329/URAT1/access/v1sub.f
+
+v1dump: An executable compiled at NERSC via:
```
You probably didn't intend to add the compiled v1dump file to git.  Please remove it.
(Adding this comment to the README, where it is good to mention the executable, since I can't seem to github comment directly on a binary file itself).

geordie666 replied:
Ooops, sorry. Should I blast that with the BFG?

sbailey replied:
It's only 30k so not super critical to blast with the BFG, but if you are comfortable with the BFG and want to use it, go for it.

geordie666 replied:
In the end I decided to be lazy as it's such a small file and will never make it into master. I'll blast it down the road if I feel energetic.

geordie666 replied:
Oh, I did remove it though.

geordie666 commented:
Note that it will now be required to have the URAT catalog to run GFAs. But, it was already required to have the suitably-formatted Gaia catalog to run GFAs. The second (Gaia) is a far greater burden that the first (URAT), so I'm comfortable with it being a requirement.

geordie666 commented:
I'd be 100% against moving the Fortran directory to the top level. I don't want to give the impression that we're expecting to add more Fortran code. In other words, Fortran is a "special case" under Python, here. The Python is a wrapper on it. Moving the Fortran to a top-level directory would imply that it has equal and substantial weight.

I also don't want to move everything into a urat directory, which would beg the question of why the `gaiamatch` code is in the `py` directory but the `uratmatch` code is in `py/urat/`. That seems structurally weird, to me.

geordie666 commented:
@sbailey: I believe I addressed all of your requests. I also: 
- moved the fortran code to `etc/fortran` as I liked that suggestion best (I've just noticed you actually suggested `etc/urat`. Let me know if you feel strongly about that distinction.
- added a `--nourat` command line argument in the binary, so if people *really* don't want to run without URAT, they don't have to.

I'm running a few tests of these changes. I'll let you know when I'm done.

geordie666 commented:
@sbailey: OK, I'm done with internal tests. Let me know if you want more changes.

sbailey commented:
Looks good; thanks.  Merging now.
